text/0000-allow-if-let-irrefutables.md

```diff
@@ -0,0 +1,78 @@
+- Feature Name: allow_if_let_irrefutables
+- Start Date: 2017-07-27
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Currently when using an if let statement and an irrefutable pattern (read always match) is used the compiler complains with an `E0162: irrefutable if-let pattern`.
+The current state breaks macros who want to accept patterns generically and this RFC proposes changing this error to an error-by-default which is allowed to be disabled by such macros.
+
+# Motivation
+[motivation]: #motivation
+
+The use cases for this is in the creation of macros where patterns are allowed because to support the `_` patterns the code has to be rewritten to be both much larger and include a compiler allow.
+The expected outcome is for irrefutable patterns to be compiled to a tautology and have the if block accept it as if it was `if true { }`.
+To support this, currently you must do something roughly the following, which seems to counteract the benefit of having if-let and while-let in the spec.
+
+```rust
+if let $p = $val {
+    $b
+}
+```
+Cannot be used, so the original match must be. The `allow` is forced so that the warning does not appear to the user of it since `_` won't be matched if `$p` is irrefutable.
+
+```rust
+#[allow(unreachable_patterns)]
```

Comment:
Question and/or Presentation nit: Is this `match $val` rust code block in the markdown meant to be associated with the above phrase "currently you must do something roughly the following," (sic) ? Its hard to understand the current presentation, because you say "roughly (like?) the following", but then the immediately following code block is the one using `if let`.

If my inference is correct in that the "the following" is meant to denote the `match $val` Rust code block, I would recommend moving the `match $val` rust code block up above the `if let` one, and add a little bit of text between the two to make their distinct roles clear.

For example:

> To support this, currently you must do something roughly the following:
> \[\[`match $val` Rust code block here\]\]
> This seems to counteract the benefit of having if-let and while-let in the spec.
> The presence of `#[allow(unreachable_patterns)]` is forced so that the warning does not appear to the user of it since `_` won't be matched if `$p` is irrefutable.
> 
> In short, one cannot expand to the following (more desirable) `if let` based code when `$p` is irrefutable.
> \[\[`if let $p` Rust code block here.\]\]
> This is what this RFC is proposing to fix, by making the compiler not as strict.